engineapi/proxy.go

```diff
@@ -1,168 +1,429 @@
+				engineClientProxy := client.getEngineClientProxy()
+				if engineClientProxy.getInUse() >= MaxProxyClientConcurrentInUse {
+					continue
+				}
+
+				exceedMaxConnection = false
+				break
+			}
+			if len(pool.clients) != 0 && exceedMaxConnection {
+				log := getLoggerForProxyHandler(h.logger, im)
+				log.Debugf("Exceed engine client proxy max concurrent use limit:%v", MaxProxyClientConcurrentInUse)
+
+				if err := h.NewEngineClientProxy(im); err != nil {
+					log.WithError(err).Debug("Failed to add more engine client proxy")
+				}
+			}
+		}
+	}
+}
+
+func (h *EngineClientProxyHandler) garbageCollectEngineClientProxy(instanceManagerImage string, clientUUID string) {
+	client, exist := h.getEngineClientProxy(instanceManagerImage, clientUUID)
+	if !exist {
+		h.logger.Debugf("Engine client proxy(%v) for %v is already gone", clientUUID, instanceManagerImage)
+		return
+	}
+
+	log := client.getLogger()
+
+	nextGC := client.getNextGarbageCollect()
+	now := time.Now()
+	if nextGC == nil || now.Before(*nextGC) {
+		return
+	}
+
+	retryFailed := ""
+	if client.getInUse() != 0 {
+		retryFailed = fmt.Sprintf("Cannot stop engine client proxy with %v still in use", client.getInUse())
+		if client.getGarbageCollectRetry() < ProxyGCMaxRetry {
+			log.Warnf(retryFailed)
+			client.setNextGarbageCollect()
+			return
+		}
+	}
+
+	log.Debug("Stopping engine client proxy")
+	if err := client.stop(); err != nil {
+		retryFailed = "Failed to stop engine client proxy"
+		if client.getGarbageCollectRetry() < ProxyGCMaxRetry {
+			log.WithError(err).Warnf(retryFailed)
+			client.setNextGarbageCollect()
+			return
+		}
+	}
+
+	if retryFailed != "" {
+		log.Warn("Exceed max retry limit, force delete engine client proxy")
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	delete(h.proxyPools[instanceManagerImage].clients, clientUUID)
+	log.Debug("Deleted engine client proxy")
+}
+
+func (h *EngineClientProxyHandler) getNewClientUUID(clients map[string]EngineClientProxy) (string, error) {
+	if len(clients) >= MaxProxyClientPerImage {
+		return "", errors.Errorf("exceed max engine client proxy per image limit: %v", MaxProxyClientPerImage)
+	}
+
+	canidateID := util.UUID()
+	if _, exist := clients[canidateID]; exist {
+		return "", errors.Errorf("proxy(%v) already exist", canidateID)
+	}
+
+	return canidateID, nil
+}
+
+func (h *EngineClientProxyHandler) getHandledEngineClientProxyPools() map[string]engineClientProxyPool {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	return h.proxyPools
+}
+
+func (h *EngineClientProxyHandler) getHandledEngineClientProxyPoolByImage(name string) (engineClientProxyPool, bool) {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	pool, exist := h.proxyPools[name]
+	return pool, exist
+}
+
+func (h *EngineClientProxyHandler) getEngineClientProxy(instanceManagerImage string, uuid string) (engineClientProxy, bool) {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	client, exist := h.proxyPools[instanceManagerImage].clients[uuid]
```

Comment:
For this function, should it check if `h.proxyPools[instanceManagerImage] != nil` first?